### PR TITLE
Add marko-web plugin package for IdentityX

### PR DESCRIPTION
**Pull request by solocommand**
Recreate #239 
- Creates the `marko-web-identity-x` package, containing dynamic Vue imports (with EventBus injection), Marko wrapping templates, and Express middleware/routing to support interacting with IdentityX.

----

# BaseCMS Marko Wrapper for IdentityX

## Installation

1. Include `@base-cms/marko-web-identity-x` as a project/website dependency.

2. Include IdentityX tenant configuration within your site
```js
// your-site/config/identity-x.js
const IdentityX = require('@base-cms/marko-web-identity-x/config');

const config = new IdentityX('<MY-APPLICATION-ID>');
module.exports = config;
```

3. Create an IdentityX router to load the IdentityX middleware.
```js
// your-site/server/routes/identity-x.js
const IdentityX = require('@base-cms/marko-web-identity-x/router');
const IdentityXConfig = require('../../config/identity-x');
const authenticate = require('../templates/user/authenticate');
const login = require('../templates/user/login');
const logout = require('../templates/user/logout');
const register = require('../templates/user/register');

module.exports = (app) => {
  IdentityX(app, IdentityXConfig);

  app.get('/user/authenticate', (req, res) => {
    res.marko(authenticate);
  });

  app.get('/user/login', (req, res) => {
    res.marko(login);
  });

  app.get('/user/logout', (req, res) => {
    res.marko(logout);
  });

  app.get('/user/register', (req, res) => {
    res.marko(register);
  });
};
```

4. Include the IdentityX router **before all other routes!**
```js
// your-site/server/routes/index.js
const IdentityX = require('./identity-x');

module.exports = (app) => {
  IdentityX(app);
  // ...
};
```

5. Create `login`, `logout`, `authenticate`, and `register` templates. These templates must include the relevant `<marko-web-identity-x-form-...>` component.
```marko
<marko-web-default-page-layout>
  <@page>
    <marko-web-identity-x-form-authenticate />
  </@page>
</marko-web-default-page-layout>
```

5. Include the Browser plugin.
```js
// your-site/browser/index.js
import IdentityX from '@base-cms/marko-web-identity-x/browser';

IdentityX(Browser);
// ...

export default Browser;
```
<!--
5. Include the styles.
```scss
// your-site/server/styles/index.scss
@import "../../node_modules/@base-cms/marko-web-identity-x/scss/form";
``` -->

## Usage

Include the `<marko-web-identity-x-form-authenticate>` component in the template where users land after authenticating (/user/authenticate).

Include the `<marko-web-identity-x-form-login>` component to display the login form.

Include the `<marko-web-identity-x-form-register>` component to display the register form.

Include the `<marko-web-identity-x-form-logout>` component to display the logout form.

Include the `<marko-web-identity-x-context>` component where you'd like access to IdentityX context.
```marko
<!-- your-site/server/templates/some-page.marko -->
<marko-web-identity-x-context|{ user, hasUser }|>
  <if(hasUser)>
    <h1>Hello ${user.givenName}!</h1>
  </if>
</identity-x>
```

Include the `<marko-web-identity-x-access>` component where you'd like to ensure access levels are met:
```marko
<!-- your-site/server/templates/content/index.marko -->
$ const { isRequired, accessLevels } = getAsObject(content, 'userRegistration');
<marko-web-identity-x-access|context|
  enabled=isRequired
  required-access-level-ids=accessLevels
>
  $ const {
    canAccess,
    isLoggedIn,
    requiresAccessLevel,
    hasRequiredAccessLevel,
    messages,
  } = context;
  <if(!canAccess)>
    <if(isLoggedIn && !hasRequiredAccessLevel)>
      $!{messages.loggedInNoAccess}
    </if>
    <else-if(!isLoggedIn && requiresAccessLevel)>
      $!{messages.loggedOutNoAccess}
    </else-if>
    <else-if(!isLoggedIn)>
      <h5>You must be logged-in to access this content.</h5>
      <cms-browser-component name="IdentitySignInForm" />
    </else-if>
  </if>
  <else>
    <p>This is secret content only some can see!</p>
  </else>
</marko-web-identity-x-access>
```

## Customization

The IdentityX form handles both login and registration based on a supplied `context` parameter. You can change the form by altering the component loader in your site's browser config to use your own Vue component:
```diff
import IdentityX from '@base-cms/marko-web-identity-x/browser';
+ import MyFormComponent from './my-form-component.vue';

-IdentityX(Browser);
+IdentityX(Browser, MyFormComponent);
```

### CODE REVIEW

File: packages/marko-web-identity-x/config.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class IdentityXConfiguration {
+  /**
+   *
+   * @param {string} appId
+   * @param {string} sentryDsn
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/config.js]: remove

File: packages/marko-web-identity-x/README.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+module.exports = config;
+```
+
+3. Include the IdentityX router **before all other routes!**
+```js
+// your-site/server/routes/index.js
+const IdentityX = require('./identity-x');
+
+module.exports = (app) => {
+  IdentityX(app);
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/README.md]: It looks like the route instructions are including twice

File: packages/marko-web-identity-x/browser/index.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const IdentityXAuthenticate = () => import(/* webpackChunkName: "identity-x-authenticate" */ '@base-cms/identity-x/browser/authenticate.vue');
+const IdentityXLogout = () => import(/* webpackChunkName: "identity-x-logout" */ '@base-cms/identity-x/browser/logout.vue');
+const IdentityXForm = () => import(/* webpackChunkName: "identity-x-form" */ '@base-cms/identity-x/browser/form.vue');
+
+export default (Browser) => {
+  const { EventBus } = Browser;
+  Browser.register('IdentityXAuthenticate', IdentityXAuthenticate, { on: { action: EventBus } });
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/browser/index.js]: How are the actions named? They'll likely need to be prefixed with `identity-x-`

File: packages/marko-web-identity-x/browser/index.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  Browser.register('IdentityXLogout', IdentityXLogout, { on: { action: EventBus } });
+  Browser.register('IdentityXForm', IdentityXForm, { on: { action: EventBus } });
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/browser/index.js]: How can a website override these forms?

File: packages/marko-web-identity-x/components/form-login.marko [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<marko-web-identity-x-context|{ user }|>
+  <marko-web-browser-component name="IdentityXForm" props={ activeUser: user, context: 'login' } />
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/components/form-login.marko]: How can a website override these forms?

File: packages/marko-web-identity-x/README.md (hunk 2 above)
Comment (solocommand) [reply packages/marko-web-identity-x/README.md]: They're not, I'm instructing the user to create an `./identity-x.js` route and include it in their `index.js` route. It's just named the same, but if anyone actually sets this up, they're just going to copy+paste anyway :P

File: packages/marko-web-identity-x/components/form-login.marko (hunk 5 above)
Comment (solocommand) [reply packages/marko-web-identity-x/components/form-login.marko]: Added ability to customize and instructions to Readme

File: packages/marko-web-identity-x/README.md (hunk 2 above)
Comment (solocommand) [reply packages/marko-web-identity-x/README.md]: Updated docs to be more clear

File: packages/marko-web-identity-x/components/form-login.marko (hunk 5 above)
Comment (solocommand) [reply packages/marko-web-identity-x/components/form-login.marko]: Added ability to customize and instructions to Readme

File: packages/marko-web-identity-x/components/context.marko [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { getAsArray } from '@base-cms/object-path';
+
+$ const { req, onAsyncBlockError } = out.global;
+$ const isDevelopment = process.env.NODE_ENV === 'development';
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/components/context.marko]: This should use the `isDev` util found in `@base-cms/marko-web/utils/is-dev`

File: packages/marko-web-identity-x/README.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+```js
+// your-site/browser/index.js
+import IdentityX from '@base-cms/marko-web-identity-x/browser';
+
+IdentityX(Browser);
+// ...
+
+export default Browser;
+```
+<!--
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/README.md]: This file is not included.

File: packages/marko-web-identity-x/browser/form/fields/email.vue [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<template>
+  <form-group>
+    <label :for="id">{{ label }}</label>
+    <input
+      :id="id"
+      v-model="email"
+      class="form-control"
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/browser/form/fields/email.vue]: Bootstrap styles... or are these supposed to be included in the `form.scss` file? (which is currently missing)

**Pull request merged**